### ISSUE TITLE
Using normal vectors for cls/seg

### ISSUE THREAD

**Issue by Crane-YU** [CVMI-Lab/PAConv#18]
Hi,
Thank you for sharing your code, just wonder how to modify your model to be suitable for additional norm vectors

**Comment by mutianxu** [comment]
Hi,

You can definitely add the norm vectors.

For how to **read** norm vectors from datasets such as ModelNet40, hopefully you can solve it by yourself. For reading the norm vectors from ShapeNet Part, we have defined it in our code.

Here I will just clarify how to use it in our model:
a. Directly concat the norm vectors with the input, making x [here](https://github.com/CVMI-Lab/PAConv/blob/main/part_seg/model/DGCNN_PAConv.py#L85) be a feature input with (3D coordinates, norm_vector), which means the input feature dim of [self.conv1](https://github.com/CVMI-Lab/PAConv/blob/main/part_seg/model/DGCNN_PAConv.py#L61) becomes (3+1)*2=8 instead of 3*2=6 before.
b. Same way in a can be also used to process ScoreNet input.
c. Note that it is possible that introducing norm vector may not bring better performance, so you may try to apply it following a or b or both of them, and welcome to share your result with us if you get better performance.

Thanks,
Mutian

**Comment by Crane-YU** [comment]
Hi @mutianxu , 
Thanks for your hints. I have tried using your method on a medical dataset and it totally fails with/without normal vector information at each point.
Specifically, I have tried different epochs from 200 to 550 with step size of 50 and lr=[0.1, 0.01, 0.001, 0.0001], and with different combinations, I still failed the classification task on a 3D medical dataset.
Just want to check if there is any unseen limitation of your methods.

**Comment by mutianxu** [comment]
Dear @Crane-YU ,

What do you mean about "fail"? Low accuracy?

Since there may be a large gap from 3D shape and 3D medical dataset, so I may recommend you to compare the difference and try to figure this out from the **data**. 

For the model, what I can make sure is that the benefits of our model illustrated in the paper all make sense. You may try to check the score learned from ScorNet, since the large gap across datasets may influence the learning of the model, such as the learned score is uniform instead of distributed.

Here are some tips you may try:
1. Delete data augmentations, or try to adapt the data pre-processing strategy used in medical task into our code.
2. Use the simplest **coordinate difference** (xj-xi, yj-yi, zj-zi) as the ScoreNet input.

Hopefully this is helpful to you.

**Comment by Crane-YU** [comment]
Hi @mutianxu ,
By "fail" I mean the loss is not decreasing after the 20th epoch and testing accuracy remains unchanged for the remaining hundreds of epochs during classification.
I am doing a binary classification task, and your cls models (both PointNet++ and DGCNN based models) fail.
However, your model works for my binary segmentation task on the same dataset.
I will try modifying your model based on the hints, and report my results here.

**Comment by Crane-YU** [comment]
> Hi @mutianxu ,
> By "fail" I mean the loss is not decreasing after the 20th epoch and testing accuracy remains unchanged for the remaining hundreds of epochs during classification.
> I am doing a binary classification task, and your cls models (both PointNet++ and DGCNN based models) fail.
> However, your model works for my binary segmentation task on the same dataset.
> I will try modifying your model based on the hints, and report my results here.

Besides, I am guesssing your model may be affected by the size of dataset?
My one only contains around 2000 samples in total for training, and few hunderds for testing.

**Comment by mutianxu** [comment]
If the model works on segmentation, then it can not be the problem of the model. Maybe you can check your classification code once again. For instance, try to use the open-source code of other models such as DGCNN of PointNet without introducing PAConv, to figure out if the problem is caused by the model or the training setting.

For the size of the dataset, I'm sorry that I can not give you a clear answer.

**Comment by Crane-YU** [comment]
@mutianxu 
I have tried my model on your another work GDANet, and some other typical works (PointConv, PointCNN, PointNet++), and for classification my dataloader works and they all give me good results.

**Comment by mutianxu** [comment]
That is a little bit weird. So far the only recommendations I can give to you are listed above. Please check again if you transfer the code of model correctly.

**Comment by Crane-YU** [comment]
Hi @mutianxu,
I have solved the issue by reducing `m` to 1 for small-scale datasets.

**Comment by mutianxu** [comment]
Well, sometimes if the dataset is too small, it will be very difficult for PAConv to learn a distributed score.

However, if you reduce m to 1, then it will be same with pure MLP.

I may recommend you to try larger m such as 2, 4.
